File: 1.HTTP-AND-CORS/code/typescript/wiring_methods_to_handlers_3.ts
import express, { Request, Response } from 'express';

const app = express();

// Express router binds method + path. Unmatched routes auto-return 404 (or can be configured for 405).
app.get('/notes', listNotes);            // safe, cacheable read
app.get('/notes/:id', getNote);
app.head('/notes/:id', getNote);         // reuse GET; framework drops the body
app.post('/notes', createNote);          // create (server assigns id)
app.put('/notes/:id', putNote);          // full replace (idempotent)
app.patch('/notes/:id', patchNote);      // partial update
app.delete('/notes/:id', deleteNote);    // remove (idempotent)

const db = {
    find: async (id: string) => { 
        if (id !== "1") throw new Error("ErrNotFound");
        return { id, content: "Note 1" }; 
    }
};

async function getNote(req: Request, res: Response) {
    const id = req.params.id;                    // built-in path params
    try {
        const note = await db.find(id);
        res.setHeader('Content-Type', 'application/json');
        res.status(200).json(note);              // 200
    } catch (err: any) {
        if (err.message === "ErrNotFound") {
            res.status(404).send('Not Found');   // 404
            return;
        }
        res.status(500).send('Server Error');
    }
}

// Dummy functions to satisfy the router
function listNotes(req: Request, res: Response) {}
function createNote(req: Request, res: Response) {}
function putNote(req: Request, res: Response) {}
function patchNote(req: Request, res: Response) {}
function deleteNote(req: Request, res: Response) {}

File: 1.HTTP-AND-CORS/code/rust/wiring_methods_to_handlers_3.rs
use actix_web::{web, App, HttpResponse, HttpServer, Responder};
use serde::Serialize;

#[derive(Serialize)]
struct Note {
    id: String,
    content: String,
}

// Actix-web binds method + path. An unmatched method auto-returns 405 Method Not Allowed.
/*
App::new()
    .route("/notes", web::get().to(list_notes))      // safe, cacheable read
    .route("/notes/{id}", web::get().to(get_note))
    .route("/notes/{id}", web::head().to(get_note))  // reuse GET; framework drops the body
    .route("/notes", web::post().to(create_note))    // create (server assigns id)
    .route("/notes/{id}", web::put().to(put_note))   // full replace (idempotent)
    .route("/notes/{id}", web::patch().to(patch_note)) // partial update
    .route("/notes/{id}", web::delete().to(delete_note)) // remove (idempotent)
*/

async fn get_note(id: web::Path<String>) -> impl Responder {
    let id_val = id.into_inner();                // built-in path params
    
    // simulated db.Find
    if id_val != "1" {
        return HttpResponse::NotFound().finish(); // 404
    }
    
    let note = Note {
        id: id_val,
        content: "Note 1".to_string(),
    };
    
    HttpResponse::Ok()                           // 200
        .content_type("application/json")
        .json(note)
}
